feat: add `math/base/special/spencef`

anandkaranubc opened the pull request:
Progresses https://github.com/stdlib-js/stdlib/issues/649

## Description

> What is the purpose of this pull request?

This pull request:

-  Adds the complete implementation for `math/base/special/spencef`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   Progresses https://github.com/stdlib-js/stdlib/issues/649

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
feat: add `math/base/special/spencef`

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: passed
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

anandkaranubc committed:
test: increase test tolerance

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
test: add tests to achieve 100% code coverage

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: passed
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
Merge branch 'develop' into feat/spencef

anandkaranubc committed:
chore: use f32 instead of float64ToFloat32

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented:
Note: This seems like the latest implementation for [`spencef`](https://github.com/scipy/xsf/blob/6062500bd11107071a62e1abba0ac02b38c9507b/include/xsf/cephes/spence.h). Confirmed it with SciPy/xsf and referenced from `deps/cephes`.

anandkaranubc committed:
Merge branch 'develop' into feat/spencef

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/spencef/lib/main.js:
```diff
@@ -0,0 +1,130 @@
+		flg |= 2;
+	}
+	if ( x > 1.5 ) {
+		w = f32( 1.0 / f32( x ) ) - 1.0;
```
Do I need to do another `f32` operation on the whole expression?

kgryte replied:
Yes.

anandkaranubc replied:
Thanks, @kgryte!

Also, in these instances, do numbers like `1.0` and `2.0`, which remain the same after the operation, also need to be wrapped for compiler reasons?

kgryte replied:
E.g., `return 1.0`?

For these, I would suggest doing something like

```
var ONE = f32( 1.0 );
var ZERO = f32( 0.0 );

// ...

function foo() {
	// ...
	return ONE;
}

// ...
```

kgryte replied:
Also means that the expression above would be

```
w = f32( f32( ONE / f32( x ) ) - ONE );
```

anandkaranubc replied:
Thanks @kgryte! I'm done with the changes. Fingers crossed.

anandkaranubc committed:
Merge branch 'develop' into feat/spencef

anandkaranubc committed:
refactor: fix `f32` emulation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
fix: use correct polyval functions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
Merge branch 'develop' into feat/spencef

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/spencef/lib/main.js:
```diff
@@ -0,0 +1,133 @@
+		w = f32( -x );
+		flg |= 1;
+	} else {
+		w = x - ONE;
```
Not sure it matters since at least the polyval functions all cast to `float32`, but to be explicit might do:

```suggestion
		w = f32( x - ONE );
```

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/spencef/lib/main.js:
```diff
@@ -0,0 +1,133 @@
+	}
+	if ( flg & 2 ) {
+		z = lnf( x );
+		y = -f32( f32( HALF*z*z ) + y );
```
```suggestion
		y = f32( -f32( f32( HALF*z*z ) + y ) );
```

Planeshifter reviewed (dismissed):
Looks overall good to me, thank you!

kgryte commented:
The following

```
y = f32( -f32( f32( HALF*z*z ) + y ) );
```

should be

```
y = f32( -f32( f32( HALF * f32( z * z ) ) + y ) );
```

anandkaranubc committed:
Merge branch 'develop' into feat/spencef

anandkaranubc committed:
chore: fix f32 emulation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented:
Thanks @Planeshifter and @kgryte for the review!

anandkaranubc committed:
Merge branch 'develop' into feat/spencef

kgryte commented on lib/node_modules/@stdlib/math/base/special/spencef/lib/main.js:
```diff
@@ -0,0 +1,133 @@
+	}
+	y = f32( -w * f32( polyvalA( w ) / polyvalB( w ) ) );
+	if ( flg & 1 ) {
+		y = f32( PI2O6 - f32( lnf( x )*lnf( ONE-x ) ) - y );
```
```suggestion
		y = f32( f32( PI2O6 - f32( lnf( x )*lnf( ONE-x ) ) ) - y );
```

kgryte committed:
refactor: cast intermediate to float32

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/math/base/special/spencef/src/main.c:
```diff
@@ -0,0 +1,148 @@
+*
+* ## Notice
+*
+* The original C code, long comment, copyright, license, and constants are from [Cephes]{@link http://www.netlib.org/cephes}. The implementation follows the original, but has been modified for JavaScript.
```
```suggestion
* The original C code, long comment, copyright, license, and constants are from [Cephes]{@link http://www.netlib.org/cephes}. The implementation follows the original, but has been modified according to stdlib conventions.
```

kgryte committed:
docs: update license comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
